PackageExplorer/Controls/MultiSelectTreeView.cs

```diff
@@ -0,0 +1,333 @@
+            {
+                var selectedItems = GetSelectedItems(treeView);
+
+                if (selectedItems != null)
+                {
+                    if (GetIsItemSelected(treeViewItem))
+                    {
+                        if (selectedItems.Contains(treeViewItem.DataContext))
+                        {
+                            return;
+                        }
+                        selectedItems.Add(treeViewItem.DataContext);
+                    }
+                    else
+                    {
+                        selectedItems.Remove(treeViewItem.DataContext);
+                    }
+                }
+            }
+        }
+
+        public static readonly DependencyProperty SelectedItemsProperty = DependencyProperty.RegisterAttached("SelectedItems", typeof(IList), typeof(MultiSelectTreeView));
+
+        public static IList GetSelectedItems(TreeView element)
+        {
+            if (element != null)
+            {
+                return (IList)element.GetValue(SelectedItemsProperty);
+            }
+            else
+            {
+                throw new ArgumentException("TreeView element is null. Cannot set value for null element.");
+            }
+        }
+
+        public static void SetSelectedItems(TreeView element, IList value)
+        {
+            if (element != null)
+            {
+                element.SetValue(SelectedItemsProperty, value);
+            }
+            else
+            {
+                throw new ArgumentNullException("TreeView element is null. Cannot set value for null element.");
+            }
+        }
+
+        private static readonly DependencyProperty StartItemProperty = DependencyProperty.RegisterAttached("StartItem", typeof(TreeViewItem), typeof(MultiSelectTreeView));
+
+
+        private static TreeViewItem GetStartItem(TreeView element)
+        {
+            return (TreeViewItem)element.GetValue(StartItemProperty);
+        }
+
+        private static void SetStartItem(TreeView element, TreeViewItem? value)
+        {
+            element.SetValue(StartItemProperty, value);
+        }
+
+
+        private static void OnTreeViewItemGotFocus(object sender, RoutedEventArgs e)
+        {
+            _selectTreeViewItemOnMouseUp = null;
+
+            if (e.OriginalSource is TreeView) return;
+
+            var treeViewItem = FindTreeViewItem(e.OriginalSource as DependencyObject);
+            if (Mouse.LeftButton == MouseButtonState.Pressed && GetIsItemSelected(treeViewItem) && Keyboard.Modifiers != ModifierKeys.Control)
+            {
+                _selectTreeViewItemOnMouseUp = treeViewItem;
+                return;
+            }
+
+            var currentTime = DateTime.Now;
+
+            // This is to prevent case, when holding CONTROL, GotFocus triggers twice (not all the time), so selected item do not get unselected right after
+            if ((currentTime - _lastTime).Milliseconds < 100)
```

Comment:
This also prevents keyboard navigation.
If you hold the up arrow key or down arrow key inside the tree view then the selection lags and get's out of sync if you press the other arrow key afterwards.

How exactly are you reproducing the CONTROL, GotFocus problem? For me it seems to work fine without it.

Reply:
Really could not find exactly why is this happening. 1st time you start app, it is okay without this line of code, 2nd time (it does not have to), but basically it gets triggered twice with CONTROL down on click, which results in not (un)selecting item you clicked on. I tried it with reduced number to 5 milliseconds, which worked for me still ok. Also I did not have selection lag so. You will see if this works.

Reply:
Looks like 5ms is small enough for holding the arrow keys down without lagging.